iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+// Replaces each usage of an entry point with its original symbol name with a
+// new symbol name.
+void replaceEntryPointUses(mlir::ModuleOp moduleOp,
+                           const DenseMap<Attribute, Attribute> &replacements) {
+  for (auto funcOp : moduleOp.getOps<mlir::FuncOp>()) {
+    funcOp.walk([&](DispatchOp dispatchOp) {
+      auto it = replacements.find(dispatchOp.entry_point());
+      if (it != replacements.end()) {
+        dispatchOp.entry_pointAttr(it->second.cast<SymbolRefAttr>());
+      }
+    });
+  }
+}
+
+bool areRegionsEquivalent(Region *lhs, Region *rhs) {
+  if (lhs->getBlocks().size() != rhs->getBlocks().size()) {
+    return false;
+  }
+
+  auto blockPairs = llvm::zip(lhs->getBlocks(), rhs->getBlocks());
+  for (auto blockPair : blockPairs) {
+    auto &lhsBlock = std::get<0>(blockPair);
+    auto &rhsBlock = std::get<1>(blockPair);
+    if (lhsBlock.getOperations().size() != rhsBlock.getOperations().size()) {
+      return false;
+    }
+
+    auto opPairs =
+        llvm::zip(lhsBlock.getOperations(), rhsBlock.getOperations());
+    for (auto opPair : opPairs) {
```

Comment:
nit: keep this in the for() - having iterator ranges as locals is weird (as they are easy to accidentally use when invalidated/etc)